### PR TITLE
Fix fakes with laravel casts

### PR DESCRIPTION
**Pull request by OliverZiegler**
This PR should fix the issue where fake fields got double encoded while creating or updating models when the field was also specified in the laravel models `$casts` property.

This is currently only tested with cast type 'object' (as this is the only interesting one for fake fields)

### CODE REVIEW

File: tests/Unit/CrudPanel/CrudPanelFakeFieldsTest.php [hunk 1]
```diff
@@ -67,14 +101,26 @@ class CrudPanelFakeFieldsTest extends BaseDBCrudPanelTest
...
 
     public function testCompactFakeFieldsFromCreateForm()
     {
         $this->crudPanel->addFields($this->fakeFieldsArray);
 
-        $compactedFakeFields = $this->crudPanel->compactFakeFields($this->fakeFieldsInputData, 'create');
+        $this->crudPanel->setModel(Article::class);
 
+        $compactedFakeFields = $this->crudPanel->compactFakeFields($this->fakeFieldsInputData, 'create');
+        dump($compactedFakeFields);
```
Comment (lloy0076) [inline tests/Unit/CrudPanel/CrudPanelFakeFieldsTest.php]: Should we dump the `dump`?

Comment (OliverZiegler) [reply tests/Unit/CrudPanel/CrudPanelFakeFieldsTest.php]: sure ;)

**Comment by tabacitu** [comment]
OMG @OliverZiegler !!! The PR also includes tests?! That's a first, I feel like you're spoiling me, I like it :-) 

Thanks a lot, wonderful PR! Here, have a GIF:

c273bdc3-b71c-4286-8449-a72ba56fbd74

**Comment by kiddtang** [comment]
@OliverZiegler  your PR makes the JSON data store in the correct form in DB.
Somehow, i need to set the ` json_decode` 2nd parameter to `true` in order to show the stored data on the fields...
https://github.com/Laravel-Backpack/CRUD/blob/4c0b8d3835a8a6742d62d466649871aed717996b/src/CrudTrait.php#L79-L81

```
$column_contents = json_decode($this->{$column}, true);   // obtain associative arrays
```

FYI, I am using v3.3.3. In my model, i just declared the `$fakeColumns` and without the `$casts`
```
protected $fakeColumns = ['fake1', 'fake2', 'fake3', 'fake4', 'contact'];
```

**Comment by OliverZiegler** [comment]
@kiddtang there is another PR for the issue you describe... I’m currently not at home to find out which

**Comment by tabacitu** [comment]
@kiddtang the PR that @OliverZiegler is talking about has been merged, a ```composer update``` should fix it for you.

**Comment by tabacitu** [comment]
@OliverZiegler , can you please help me replicate the problem to test if this fixes it? I can't seem to get the same issue on ```CRUD 3.3.4```, no matter what I try.

I've tried:
- a fake ```table``` field in Monster ([video here](https://cl.ly/06152B1d0v05));
- a regular ```table``` field in Monster ([video here](https://cl.ly/2i1z392X1u0P));
- a regular translatable ```table``` field in Product ([video here](https://cl.ly/410Q130T1u0I));
- a fake translatable ```table``` field in Product ([video here](https://cl.ly/1X1x191k1Z2B));

All cast as object; and non-casted. I think I've tried all possible use cases - but none seem to double-encode the data.... All seem to work fine for me...

How did you get the double-enconding problem?

Thanks, cheers!

**Comment by kiddtang** [comment]
@tabacitu actually i have commented in another PR https://github.com/Laravel-Backpack/CRUD/pull/687#issuecomment-352321329
A few PR for fake fields... which one you're referring to?
I have tested with **CRUD 3.3.3** without the @OliverZiegler 's PR #1116
(Note: doing php artisan cache:clear before each modification of `CrudTrait.php`), i observe no double encoded.
However, without cache clearing, somehow double encoding can be happened again. i am not sure why.

I will test CRUD 3.3.4 again with my project. Thanks for the fixing!

**Comment by OliverZiegler** [comment]
@tabacitu we got this issue in several projects and it happens always.

I'm not sure why there is such an issue merging and closing this PR.
Just try the provided tests without the two (minor) changes I made.
Just returning `true` in `shouldEncodeFakes` would also be sufficient.

The tests will fail without my changes.

```
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
-    'cast_metas' => Array (...)
-    'cast_tags' => Array (...)
-    'cast_extras' => Array (...)
+    'cast_metas' => '{"cast_meta_title":"Meta Titl...alue"}'
+    'cast_tags' => '{"cast_tags":["tag1","tag2","tag3"]}'
+    'cast_extras' => '{"cast_extra_details":["detai...il3"]}'
```

I currently installed CRUD 3.3.5 and this is still an issue.
In my current project I have following setup:

I got a model with 

```php
protected $casts = [
    'extras'   => 'object',
];
```

and in CrudController I have

```php
$this->crud->addFields([
            [
                'name' => 'firstname',
                'label' => 'Vorname',
                'type' => 'text',
                'fake' => true,
                'store_in' => 'extras'
            ],
            [
                'name' => 'lastname',
                'label' => 'Nachname',
                'type' => 'text',
                'fake' => true,
                'store_in' => 'extras'
            ],
])
```

The edit form shows the data (as it gets double_decoded, once by laravel and once by backpack) but in the table json is double encoded after save (cause backpack encodes and then the laravel cast encodes).

**Comment by tswonke** [comment]
I really hope this is going to be merged soon. Just as Oli wrote we are using fake fields in several projects

**Comment by tswonke** [comment]
Any update here? :)

**Comment by OliverZiegler** [comment]
People? What's up with this? 
I just started another project still having this issue with backpack/crud 3.3.6...

It's kind of frustrating creating PRs **with Tests** to properly confirm the issue and that it is resolved and waiting for the merge over 2 months now...

**Comment by OliverZiegler** [comment]
@tabacitu what's wrong with this one? Still waiting for the merge... as this is still an issue in pur projects...

**Comment by telkins** [comment]
@tabacitu I'd like to chime in, too, if I may.  It's not 100% clear to me, but it *seems* like some are saying that this is resolved and some are saying it's not.  Well, I'm using CRUD 3.3.11, I have a simple model with one JSON field, `links`.  I cast `links` to `array` and I add it to `fakeColumns`.  It seems `CrudTrait` tries to `json_decode` it....blowing up in the process.

I can remove the cast, but then I lose elsewhere.  :-(

I hope that something can be done about this and I'm happy to help out if needed.  Please just let me know what needs to be done.  :-)

**Comment by telkins** [comment]
@tabacitu I should add that it seems that the changes in this PR do not address the specific problem I'm experiencing, which is occurring in `CrudTrait` when it tries to decode the JSON...unnecessarily:
```
    public function addFakes($columns = ['extras'])
    {
        foreach ($columns as $key => $column) {
            $column_contents = $this->{$column};

            if (! is_object($this->{$column})) {
                $column_contents = json_decode($this->{$column});
            }

            if ((is_array($column_contents) || is_object($column_contents) || $column_contents instanceof Traversable)) {
                foreach ($column_contents as $fake_field_name => $fake_field_value) {
                    $this->setAttribute($fake_field_name, $fake_field_value);
                }
            }
        }
    }
```

This should likely take advantage of the new `shouldEncodeFake` method.

**Comment by OliverZiegler** [comment]
@telkins well it seems you found another bug with fake_fields. 
As mentioned above I only address the issue when casting to `object`.

Will try to add some failing tests for your issue. 
Perhaps people will then see our problems with this. 
I did add tests here (that fail without my fixes) but it seems no one is really interested in failing tests and fixing PRs these days. 🤔 😉 

The fix for casting to `array` could then be:

```php
public function addFakes($columns = ['extras'])
    {
        foreach ($columns as $key => $column) {
            $column_contents = $this->{$column};

            if (! is_object($this->{$column}) && ! $this->shouldEncodeFake($column)) {
                $column_contents = json_decode($this->{$column});
            }

            if ((is_array($column_contents) || is_object($column_contents) || $column_contents instanceof Traversable)) {
                foreach ($column_contents as $fake_field_name => $fake_field_value) {
                    $this->setAttribute($fake_field_name, $fake_field_value);
                }
            }
        }
    }
```

but need to verify this with proper unit tests.

**Comment by telkins** [comment]
@OliverZiegler Thanks for the quick feedback.

It was a bit hard for me to be certain whether or not I was experiencing exactly the same problem.  It's definitely in the same ballpark, though.  At least that much is clear.  :-)

Before I found this PR, I changed this bit (from your post, above):
```
            if (! is_object($this->{$column}) && ! $this->shouldEncodeFake($column)) {
                $column_contents = json_decode($this->{$column});
            }
```
...to this:
```
            if (is_string($this->{$column})) {
                $column_contents = json_decode($this->{$column});
            }
```

I thought it was that simple...but perhaps it's not.  It worked for me and I didn't want to spend too much time worrying about it if someone else had already tackled it.  I don't know all the various scenarios one might encounter here, but one can't `json_decode` anything but a string....right...?  :-)

**Comment by telkins** [comment]
...of course, maybe there are times where one wouldn't want to `json_decode` `$this->{$column}` at this point even if it is a string....?  :-|

**Comment by OliverZiegler** [comment]
@telkins well from the php docs `json_decode` is defined like 

```php
mixed json_decode ( string $json [, bool $assoc = FALSE [, int $depth = 512 [, int $options = 0 ]]] )
```

so your 

```php
if (is_string($this->{$column})) {
    $column_contents = json_decode($this->{$column});
}
```
seems like the best approach to be certain this would not crash.

And when decoding fake fields there should never be the point where we have a string in `$column` but don't want to decode it. Fake fields are per definition (or better said convention?) arrays or objects. So this should be fine.

**Comment by OliverZiegler** [comment]
@tabacitu Thanks for (finally) merging 😊 

This is not an easy issue (as discussion shows) and I understand your extensive testing.
Did you test anything not currently handled by the unit tests? If yes you should create an issue to improve unit testing. 

If we could rely on the tests there is no need for you to spend 3-4 hours but you just run `phpunit` 😉

**Comment by tabacitu** [comment]
Yup. I agree, more unit tests is something we could _definitely_ use. Translatable fields is what I've focused on most - I've gone through each of them to make sure they work without casting to array / object, and used it on a few personal apps to make sure it doesn't break people's apps. I still have a slight fear that it will, though... That's exactly what unit tests should solve, I know :-)

Thanks again!

**Comment by tswonke** [comment]
Thank you @tabacitu, this helps @OliverZiegler and me/us a lot! :-)

**Pull request merged**